From 987b8403cfc614ad96f6aa8e80344224c31bf0d5 Mon Sep 17 00:00:00 2001
Message-Id: <987b8403cfc614ad96f6aa8e80344224c31bf0d5.1431491843.git.chang-joon.lee@intel.com>
From: Vidya Srinivas <vidya.srinivas@intel.com>
Date: Fri, 8 May 2015 12:05:26 +0530
Subject: [PATCH] SQUASHME! [VPG]: drm/i915: Enable primary unconditionally
 only during modeset path

Enable the primary plane without condition should be done only during the
modeset path. Else, primary plane gets enabled even when not intentional in
some scenarios where we force disable primary plane to save power.

Squash this patch to:
	commit e70ad15e2f08f1329229dc6f689f5977ea8e8438
	Author: Vidya Srinivas <vidya.srinivas@intel.com>
	Date:   Wed Apr 15 13:59:01 2015 +0530
	REVERTME [VPG]: drm/i915: Fix the plane enable sequence.
	Change-Id: I5eee0a32d3f855f8fc538e177693c7560502b5a2

Issue: GMINL-9076
Change-Id: Id1b070e04935d449391b38b4161fc89b576612f1
Signed-off-by: Arun R Murthy <arun.r.murthy@intel.com>
Signed-off-by: Vidya Srinivas <vidya.srinivas@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c |   12 +++++++-----
 drivers/gpu/drm/i915/intel_drv.h     |    1 +
 2 files changed, 8 insertions(+), 5 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 2c80741..e42dc6f 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -2346,6 +2346,7 @@ static void intel_enable_primary_hw_plane(struct drm_i915_private *dev_priv,
 
 	intel_crtc->primary_enabled = true;
 	dev_priv->pipe_plane_stat |= VLV_UPDATEPLANE_STAT_PRIM_PER_PIPE(pipe);
+	intel_crtc->enableprimary = true;
 	dev_priv->display.update_primary_plane(crtc, crtc->primary->fb,
 					       crtc->x, crtc->y);
 
@@ -2763,7 +2764,11 @@ static void i9xx_update_primary_plane(struct drm_crtc *crtc,
 		intel_crtc->pri_update = false;
 	} else {
 		dspcntr = I915_READ(reg);
-		dspcntr |= DISPLAY_PLANE_ENABLE;
+		if ((atomic_read(&dev_priv->psr.update_pending)) ||
+			(intel_crtc->enableprimary)) {
+			dspcntr |= DISPLAY_PLANE_ENABLE;
+			intel_crtc->enableprimary = false;
+		}
 	}
 
 	/*
@@ -2932,9 +2937,6 @@ static void i9xx_update_primary_plane(struct drm_crtc *crtc,
 		intel_crtc->dspaddr_offset = linear_offset;
 	}
 
-	DRM_DEBUG_KMS("Writing base %08lX %08lX %d %d %d\n",
-		      i915_gem_obj_ggtt_offset(obj), linear_offset, x, y,
-		      fb->pitches[0]);
 	intel_crtc->reg.stride = fb->pitches[0];
 	if (!dev_priv->atomic_update)
 		I915_WRITE(DSPSTRIDE(plane), intel_crtc->reg.stride);
@@ -13027,7 +13029,7 @@ static void intel_crtc_init(struct drm_device *dev, int pipe)
 	intel_crtc->rotate180 = false;
 	/* Flag for wake from sleep */
 	dev_priv->is_resuming = false;
-
+	intel_crtc->enableprimary = false;
 	WARN_ON(drm_crtc_index(&intel_crtc->base) != intel_crtc->pipe);
 }
 
diff --git a/drivers/gpu/drm/i915/intel_drv.h b/drivers/gpu/drm/i915/intel_drv.h
index 4c1ccb4..5596f7d 100644
--- a/drivers/gpu/drm/i915/intel_drv.h
+++ b/drivers/gpu/drm/i915/intel_drv.h
@@ -450,6 +450,7 @@ struct intel_crtc {
 	bool active;
 	unsigned long enabled_power_domains;
 	bool primary_enabled; /* is the primary plane (partially) visible? */
+	bool enableprimary;
 	bool lowfreq_avail;
 	bool pri_update;
 	struct intel_overlay *overlay;
-- 
1.7.9.5

